benchmark/UnsortedAndSortedJoinBenchmark.cpp

```diff
@@ -0,0 +1,92 @@
+
+// Benchmarks for unsorted and sorted tables, with and without overlapping values in
+// IdTables. Done with normal join and hash join.
+void BM_UnsortedAndSortedIdTable(BenchmarkRecords* records) {
+  // For easier changing of the IdTables size.
+  const size_t NUMBER_ROWS = 10000;
+  const size_t NUMBER_COLUMNS = 20; 
+
+  auto hashJoinLambda = makeHashJoinLambda();
+  auto joinLambda = makeJoinLambda();
+
+  // Tables, that have no overlapping values in their join columns.
+  IdTableAndJoinColumn a{
+    createRandomlyFilledIdTable(NUMBER_ROWS, NUMBER_COLUMNS, 0, 0, 10), 0};
+  IdTableAndJoinColumn b{
+    createRandomlyFilledIdTable(NUMBER_ROWS, NUMBER_COLUMNS, 0, 20, 30), 0};
+
+  // Lambda wrapper for the functions, that I measure.
+  
+  // Sorts the table IN PLACE and the uses the normal join on them.
+  auto sortThenJoinLambdaWrapper = [&]() {
+        // Sorting the tables by the join column.
+        Engine::sort<NUMBER_COLUMNS>(&a.idTable, 0);
+        Engine::sort<NUMBER_COLUMNS>(&b.idTable, 0);
+
+        useJoinFunctionOnIdTables(a, b, joinLambda);
+  };
+
+  auto joinLambdaWrapper = [&]() {
+    useJoinFunctionOnIdTables(a, b, joinLambda);
+  };
+
+  auto hashJoinLambdaWrapper = [&]() {
+        useJoinFunctionOnIdTables(a, b, hashJoinLambda);
+  };
+
+  // Because it's easier to read/interpret, the benchmarks are entries in tables.
+  records->addTable("Sorted IdTables", {"Merge join", "Hashed join"},
+      {"Overlapping join column entries", "Non-overlapping join column entries"});
+  records->addTable("Unsorted IdTables", {"Merge join", "Hashed join"},
+      {"Overlapping join column entries", "Non-overlapping join column entries"});
+
+  // Benchmarking with non-overlapping IdTables.
+ 
+  records->addToExistingTable("Unsorted IdTables", 1, 1, hashJoinLambdaWrapper);
+  records->addToExistingTable("Unsorted IdTables", 0, 1, sortThenJoinLambdaWrapper);
+
+  // Because the sortThenJoinLambdaWrapper sorts tables IN PLACE, a and b are now sorted.
+  records->addToExistingTable("Sorted IdTables", 1, 1, hashJoinLambdaWrapper);
+  records->addToExistingTable("Sorted IdTables", 0, 1, joinLambdaWrapper);
+
+  // Benchmarking with overlapping IdTables.
+
+  // We make the tables overlapping and then randomly shuffle them.
+  for (size_t i = 0; i*20 < NUMBER_ROWS; i++) {
+    a.idTable(i*10, 0) = b.idTable(i*20, 0);
+  }
+  randomShuffle(a.idTable.begin(), a.idTable.end());
+  randomShuffle(b.idTable.begin(), b.idTable.end());
+
+  records->addToExistingTable("Unsorted IdTables", 1, 0, hashJoinLambdaWrapper);
+  records->addToExistingTable("Unsorted IdTables", 0, 0, sortThenJoinLambdaWrapper);
+
+  // Because the sortThenJoinLambdaWrapper sorts tables IN PLACE, a and b are now sorted.
+  records->addToExistingTable("Sorted IdTables", 1, 0, hashJoinLambdaWrapper);
+  records->addToExistingTable("Sorted IdTables", 0, 0, joinLambdaWrapper);
+}
+
+BenchmarkRegister temp{{BM_UnsortedAndSortedIdTable}};
```

Comment:
I have deliberately not looked closely at this file. Maybe our target would be to first merge the initial benchmark suite, and then in a second step use it to get all the measures for the Join implementations.
That way each of the PRs have still manageable sizes.